added support for fields.List.serialize to correctly process generators (

aganezov opened the pull request:
following our discussion on #185 I've made small workaround changes to the fields.List.serialize method to support generators.

additional test cases were added and successfully passed, as well as the rest of the test suite

while my workaround is simple, it does not support the generator protocol in full (we create dummy one entry list every time we go through generator, and no "exception chaining", no "send()", etc is provided), but I don't think this is that important at the moment.

aganezov committed:
added support for fields.List to correctly process generators and not just non string objects with __getitem__ method
added respective test cases to the test suite

davidism commented:
Why not just take out both checks (generator and the one below it about index/string) and just allow any iterable?  This would be *way* more convenient when, for example, you have a set that you want to serialize, since there is no set in JSON and a list is probably good enough.

The one downside is that it will now accept a bare string, but that "problem" will pretty much be immediately apparent when you get a list of single characters.

sergey-aganezov-jr commented:
I don't think `string` issue will be that big, as in this case it would be a "hack" to go from string to a list of chars, by simply specifying `fields.List(fields.String())` on the string value in the serialized class. And if people want to serialize string, they shall really use `fields.String()`.

I believe this pull request is the first step, and then `fields.List` can be generalized into supporting all iterables. On the other hand, it is not that easy to check if something is `iterable` in python ([link](http://stackoverflow.com/questions/1952464/in-python-how-do-i-determine-if-an-object-is-iterable)).

It seems that something like this can be done:

```python
    if utils.is_indexable_but_not_string(value) and not isinstance(value, dict):
        return [self.container.serialize(idx, value) for idx in range(len(value))]
    if utils.is_iterable_but_not_string(value) and not isinstance(value, dict):
        return [self.container.serialize(0, [entry]) for entry in value]
    ....
````

and `utils.is_iterable_but_not_string(value)` can be implemented something like this:
```python
    def is_iterable_but_not_string(obj):
        return (hasattr(obj, "__iter__") or isinstance(obj, types.GeneratorType)) and not not hasattr(obj, "strip")

aganezov committed:
updated implementation for the fields.List.serialize method to support all iterable objects as input, as well as indexables
updated test cases to check for support of custom classes with __iter__ and set object

aganezov committed:
continuation of #185
fixed test case built-in set support from fields.List to be compatible with python 2.6 (set comprehension was removed)

aganezov committed:
continuation of #185
updated fields.List.serialize implementation to process all iterables as well as custom indexables with length
updated respective test cases

aganezov committed:
continuation of #185
removed support for the indexable objects that have not __iter__

aganezov committed:
continuation of #185
quick follow up: unused import removed
